src/module_2/module_2_part_1.md

```diff
@@ -0,0 +1,1349 @@
+      <td>1756.21</td>
+    </tr>
+    <tr>
+      <th>2260</th>
+      <td>4021cf86ce45a4fe0793a16fa6c5f8c0fc104d56cf93a4...</td>
+      <td>383</td>
+      <td>cleaning-products</td>
+      <td>6</td>
+      <td>cleaning-products</td>
+      <td>1738.17</td>
+    </tr>
+    <tr>
+      <th>1258</th>
+      <td>f49d903f33fde7436da05a2dc6eddebff62cd9ca552121...</td>
+      <td>337</td>
+      <td>toilet-roll-kitchen-roll-tissue</td>
+      <td>41</td>
+      <td>dishwashing</td>
+      <td>1617.47</td>
+    </tr>
+    <tr>
+      <th>...</th>
+      <td>...</td>
+      <td>...</td>
+      <td>...</td>
+      <td>...</td>
+      <td>...</td>
+      <td>...</td>
+    </tr>
+    <tr>
+      <th>1756</th>
+      <td>769555ec9d277b7fa4a55e7f5c00f27017afbb52cc5056...</td>
+      <td>2</td>
+      <td>tins-packaged-foods</td>
+      <td>0</td>
+      <td>NaN</td>
+      <td>2.98</td>
+    </tr>
+    <tr>
+      <th>1654</th>
+      <td>70b771db8fe2d0c96f12e1b53b472ea646f16be0f7ce34...</td>
+      <td>1</td>
+      <td>cooking-ingredients</td>
+      <td>0</td>
+      <td>NaN</td>
+      <td>2.29</td>
+    </tr>
+    <tr>
+      <th>89</th>
+      <td>864e7f2d544ae108b74798fcfc61b4cf23e16d84921f35...</td>
+      <td>1</td>
+      <td>cleaning-products</td>
+      <td>0</td>
+      <td>NaN</td>
+      <td>1.99</td>
+    </tr>
+    <tr>
+      <th>1507</th>
+      <td>2d4dca7ab4ae6e5b3fc418d971ee8ad74462d63093db04...</td>
+      <td>1</td>
+      <td>cooking-ingredients</td>
+      <td>1</td>
+      <td>cooking-ingredients</td>
+      <td>0.99</td>
+    </tr>
+    <tr>
+      <th>2458</th>
+      <td>fb0f385e8eb9595411cddee4e4aadc1de728be7c709be1...</td>
+      <td>1</td>
+      <td>skincare</td>
+      <td>0</td>
+      <td>NaN</td>
+      <td>0.99</td>
+    </tr>
+  </tbody>
+</table>
+<p>4948 rows × 6 columns</p>
+</div>
+
+
+
+As we can see, the favourite product types of the clients with highest revenues are among the most bought product types. So now, the main aspect to check is, when is the time where more carts are abandoned and what type of products belong to these carts? 
+
+
+```python
+def analyze_abandoned_carts():
+    # Convertir timestamp a hora
+    df_abandoned_carts['hour'] = df_abandoned_carts['created_at'].dt.hour
+
+    # Conteo de abandonos por hora
+    hour_counts = df_abandoned_carts['hour'].value_counts()
+
+    # Gráfico circular de abandonos por hora
+    plt.figure(figsize=(8, 8))
+    hour_counts.plot(kind='pie', autopct='%1.1f%%')
+    plt.title('Abandonos de Carritos por Hora')
+    plt.ylabel('')
+    plt.show()
+
+    # Obtener las tres horas con más abandonos
+    top_hours = hour_counts.head(3).index
+
+    # Analizar los tipos de productos más abandonados en esas horas
+    for hour in top_hours:
+        # Filtrar carritos abandonados en la hora específica
+        hour_abandoned = df_abandoned_carts[df_abandoned_carts['hour'] == hour]
+        
+        # Obtener los variant_id y luego los product_type
```

Comment:
Again, you can do this operation for the whole dataset rather than inside each hour loop.

```
hours_abandoned = df_abandoned_carts.explode('variant_id').merge(inventory_df[['variant_id', 'product_type']], on='variant_id')
```

then for each hour
```
product_type_counts = hours_abandoned.loc[lambda x: x.hour == hour].product_types.value_counts()
```